thirdparty/DiskANN/src/pq_flash_index.cpp

```diff
@@ -58,6 +58,130 @@ namespace {
 }  // namespace
 
 namespace diskann {
+  template<typename T>
+  IteratorWorkspace<T>::IteratorWorkspace(
+      const T *query_data, const diskann::Metric metric,
+      const uint64_t aligned_dim, const uint64_t data_dim, const float alpha,
+      const uint64_t lsearch, const uint64_t beam_width,
+      const float filter_ratio, const bool for_tuning,
+      const float max_base_norm, const knowhere::BitsetView &bitset)
+      : lsearch(lsearch), beam_width(beam_width), filter_ratio(filter_ratio),
+        metric(metric), alpha(alpha), for_tuning(for_tuning),
+        max_base_norm(max_base_norm), bitset(bitset) {
+    frontier.reserve(2 * beam_width);
+    frontier_nhoods.reserve(2 * beam_width);
+    frontier_read_reqs.reserve(2 * beam_width);
+    cached_nhoods.reserve(2 * beam_width);
+
+    // own query and query_T
+    diskann::alloc_aligned((void **) &aligned_query_T, aligned_dim * sizeof(T),
+                           8 * sizeof(T));
+    diskann::alloc_aligned((void **) &aligned_query_float,
+                           aligned_dim * sizeof(float), 8 * sizeof(float));
+    memset((void *) aligned_query_T, 0, aligned_dim * sizeof(T));
+    memset(aligned_query_float, 0, aligned_dim * sizeof(float));
+    q_dim = data_dim;
+    if (metric == diskann::Metric::INNER_PRODUCT) {
+      // query_dim need to be specially treated when using IP
+      q_dim--;
+    }
+    for (uint32_t i = 0; i < q_dim; i++) {
+      aligned_query_T[i] = query_data[i];
+      aligned_query_float[i] = (float) query_data[i];
+      query_norm += (float) query_data[i] * (float) query_data[i];
+    }
+
+    if (metric == diskann::Metric::INNER_PRODUCT ||
+        metric == diskann::Metric::COSINE) {
+      query_norm = std::sqrt(query_norm);
+      if (metric == diskann::Metric::INNER_PRODUCT) {
+        aligned_query_T[q_dim] = 0;
+        aligned_query_float[q_dim] = 0;
+      }
+      for (uint32_t i = 0; i < q_dim; i++) {
```

Comment:
missing `if (metric == diskann::Metric::COSINE)` ?

Reply:
There is a special calculation method for `IP` metric, and an extra dim (the last) will be used for “convert MIPS to L2 search”.